src/Microsoft.AspNetCore.Mvc.Core/ApplicationModels/ApplicationModelFactory.cs

```diff
@@ -0,0 +1,364 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using Microsoft.AspNetCore.Mvc.ActionConstraints;
+using Microsoft.AspNetCore.Mvc.Core;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Extensions.Internal;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Mvc.ApplicationModels
+{
+    /// <summary>
+    /// A facade service for creating application models.
+    /// </summary>
+    internal class ApplicationModelFactory
+    {
+        private readonly IApplicationModelProvider[] _applicationModelProviders;
+        private readonly IList<IApplicationModelConvention> _conventions;
+
+        public ApplicationModelFactory(
+            IEnumerable<IApplicationModelProvider> applicationModelProviders,
+            IOptions<MvcOptions> options)
+        {
+            if (applicationModelProviders == null)
+            {
+                throw new ArgumentNullException(nameof(applicationModelProviders));
+            }
+
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            _applicationModelProviders = applicationModelProviders.OrderBy(p => p.Order).ToArray();
+            _conventions = options.Value.Conventions;
+        }
+
+        public ApplicationModel CreateApplicationModel(IEnumerable<TypeInfo> controllerTypes)
+        {
+            if (controllerTypes == null)
+            {
+                throw new ArgumentNullException(nameof(controllerTypes));
+            }
+
+            var context = new ApplicationModelProviderContext(controllerTypes);
+
+            for (var i = 0; i < _applicationModelProviders.Length; i++)
+            {
+                _applicationModelProviders[i].OnProvidersExecuting(context);
+            }
+
+            for (var i = _applicationModelProviders.Length - 1; i >= 0; i--)
+            {
+                _applicationModelProviders[i].OnProvidersExecuted(context);
+            }
+
+            ApplicationModelConventions.ApplyConventions(context.Result, _conventions);
```

Comment:
This whole part of the codebase is a minefield of static classes with a single method and a few unit tests. I don't see it as a problem, so I tried to keep the existing stuff from changing where possible.